server/routing/routing_remote.go

```diff
@@ -179,26 +205,73 @@ func newRemote(ctx context.Context,
-	remoteLogger.Debug("Called remote routing's Publish method for network operations", "ref", ref, "record", record)
-
-	decodedCID, err := cid.Decode(ref.GetCid())
+// publishToNetwork is the core publishing logic shared by Publish and republishing tasks.
+// It announces a record's CID and labels to the network via DHT and GossipSub.
+//
+// This method captures receiver state (server, pubsubManager) and can be passed
+// as a callback to CleanupManager without creating circular dependencies.
+//
+// Flow:
+//  1. Parse and validate CID
+//  2. Announce CID to DHT (critical - returns error if fails)
+//  3. Publish labels via GossipSub (best-effort - logs warning if fails)
+//
+// Parameters:
+//   - ctx: Operation context
+//   - cidStr: Content ID as string
+//   - record: Record to extract labels from
+//
+// Returns:
+//   - error: If critical operations fail (CID parsing, DHT announcement)
+func (r *routeRemote) publishToNetwork(ctx context.Context, cidStr string, record *corev1.Record) error {
+	// Parse and validate CID
+	decodedCID, err := cid.Decode(cidStr)
 	if err != nil {
-		return status.Errorf(codes.InvalidArgument, "failed to parse CID: %v", err)
+		return fmt.Errorf("failed to parse CID: %w", err)
 	}
 
-	// Announce CID to DHT network (triggers pull-based discovery)
+	// 1. Announce CID to DHT network (content discovery)
 	err = r.server.DHT().Provide(ctx, decodedCID, true)
 	if err != nil {
-		return status.Errorf(codes.Internal, "failed to announce object %v: %v", ref.GetCid(), err)
+		return fmt.Errorf("failed to announce CID to DHT: %w", err)
+	}
+
+	// 2. Publish labels via GossipSub (if enabled)
+	// This provides efficient label propagation to ALL subscribed peers
+	if r.pubsubManager != nil {
+		labelList := GetLabelsFromRecord(record)
+		if len(labelList) > 0 {
+			if err := r.pubsubManager.PublishLabels(ctx, cidStr, labelList); err != nil {
+				// Log warning but don't fail - DHT announcement already succeeded
+				// Remote peers can still discover via DHT+Pull fallback
+				remoteLogger.Warn("Failed to publish labels via GossipSub",
+					"cid", cidStr,
+					"error", err,
+					"fallback", "DHT+Pull will handle discovery")
+			} else {
+				remoteLogger.Debug("Successfully published labels via GossipSub",
+					"cid", cidStr,
+					"labels", len(labelList),
+					"topicPeers", len(r.pubsubManager.GetTopicPeers()))
+			}
+		}
 	}
 
-	// Note: Label announcements via DHT.PutValue() have been removed.
-	// Labels are now discovered via pull-based mechanism when remote peers
-	// receive the CID provider announcement and pull the content.
+	return nil
+}
+
+func (r *routeRemote) Publish(ctx context.Context, ref *corev1.RecordRef, record *corev1.Record) error {
+	remoteLogger.Debug("Publishing record to network", "ref", ref)
+
+	// Use shared publishing logic
+	if err := r.publishToNetwork(ctx, ref.GetCid(), record); err != nil {
+		return status.Errorf(codes.Internal, "%v", err)
+	}
```

Comment:
there is no need to have 2 identical functions here, lets keep one

Reply:
fixed